polar-core/src/vm.rs

```diff
@@ -1889,25 +1969,185 @@ impl PolarVirtualMachine {
                 ))
             }
         }
-        Ok(())
+        Ok(QueryEvent::None)
     }
 
+    /// Handle variables & constraints as arguments to various operations.
+    /// Calls the `eval` method to handle ground terms.
+    ///
+    /// Arguments:
+    ///
+    /// - handle_unbound_left_var: If set to `false`, allow `eval` to handle
+    ///   operations with an unbound left variable, instead of adding a constraint.
+    ///   Some operations, like `In`, emit new goals or choice points when the left
+    ///   operand is a variable.
+    /// - handle_unbound_right_var: Same as above but for the RHS. `Dot` uses this.
     #[allow(clippy::many_single_char_names)]
-    fn in_op_helper(
+    fn query_op_helper<F>(
         &mut self,
         term: &Term,
-        op: Operator,
-        args: Vec<Term>,
-    ) -> PolarResult<QueryEvent> {
-        assert_eq!(args.len(), 2);
-        let item = &args[0];
-        let iterable = &args[1];
-        match (item.value(), iterable.value()) {
+        eval: F,
+        handle_unbound_left_var: bool,
+        handle_unbound_right_var: bool,
+    ) -> PolarResult<QueryEvent>
+    where
+        F: Fn(&mut Self, &Term) -> PolarResult<QueryEvent>,
+    {
+        let Operation { operator: op, args } = term.value().as_expression().unwrap();
+
+        let mut args = args.clone();
+        assert!(args.len() >= 2);
+        let left = &args[0];
+        let right = &args[1];
+
+        match (left.value(), right.value()) {
             (Value::Expression(_), _)
             | (_, Value::Expression(_))
             | (Value::RestVariable(_), _)
-            | (_, Value::RestVariable(_)) => unreachable!("invalid syntax"),
+            | (_, Value::RestVariable(_)) => {
+                panic!("invalid query");
+            }
+            _ => {}
+        };
+
+        if let Value::Variable(r) = right.value() {
+            // A variable on the right, ground on the left.
```

Comment:
`ground on the left` isn't assured at this point, right?

Reply:
Right; 7729e0e4.